### ISSUE TITLE
Is there any way to build IfcOpenShell to java interface or js ？

### ISSUE THREAD

**Issue by hxboxy-xh** [IfcOpenShell/IfcOpenShell#2265]
now IfcOpenShell  can build Python interface  use swig，so it can build to other language？

**Comment by aothms** [comment]
Definitely. You can use SWIG to generate a JNI (java native interface) module. For JavaScript I'd probably recommend using emscriptem to compile to WASM (web assembly) which can be imported in the browser. I don't have time or budget to work on this now, but I can provide assistance and advice if you're interested in working on this.

**Comment by TZwielehner** [comment]
Maybe ifc.js would be a more straightforward solution. Apstex for Java.

Xbim for c#, ifc++ (ifcquery) for c++.
And ifcopenshell for python / c++. Did I forget anything here? ;)

**Comment by aothms** [comment]
Programming language is really only one aspect to evaluate a software library on. Most (nearly all) mature geometry libraries are C++ (maybe rust in the near future). So if handling of complex sweeps, boolean operations with tolerance, curved surfaces is of interest wrapping c++ libraries into the language of your choice is a reasonable option. Similar thoughts could exist around how IFC geometry is handled: e.g strict or lax, early or late bound, support for serialization formats, etc.

**Comment by BIMvoice** [comment]
Closing as answered. @aothms covered the approaches in 2022 (SWIG JNI for Java, emscripten/WASM for JavaScript), and since then the JavaScript direction has materialized concretely: the build system now has a `WASM_BUILD` option, ifcopenshell ships a WebAssembly wheel usable in the browser via pyodide, and there is active web-viewer work built on the WASM build. For Java, the SWIG JNI route aothms described remains the way to go and would need a contributor to take it up (the old ifcjni component was removed long ago). If anyone wants to pursue a Java binding, a fresh dedicated issue or discussion would be the right place.

This comment was written with AI assistance.
